src/ExpressionEvaluator/Core/Source/ExpressionCompiler/MetadataUtilities.cs

```diff
@@ -105,42 +114,98 @@ internal static partial class MetadataUtilities
+                {
+                    var reader = metadata.MetadataReader;
+                    var identity = reader.ReadAssemblyIdentityOrThrow();
+                    if (referencesByIdentity != null)
+                    {
+                        string identityKey = identity.Name;
+                        ImmutableArray<(AssemblyIdentity, MetadataReference)> references;
+                        if (!referencesByIdentity.TryGetValue(identityKey, out references))
+                        {
+                            references = ImmutableArray<(AssemblyIdentity, MetadataReference)>.Empty;
+                        }
+                        referencesByIdentity[identityKey] = references.Add((identity, reference));
+                    }
+                    if (targetReference == null &&
+                        reader.GetModuleVersionIdOrThrow() == moduleVersionId)
+                    {
+                        targetReference = reference;
+                    }
+                    continue;
+                }
+
+                referencesBuilder.Add(reference);
                 if (identitiesBuilder != null)
                 {
                     var reader = metadata.MetadataReader;
                     var identity = reader.ReadAssemblyIdentityOrThrow();
                     identitiesBuilder.Add(identity);
-                    if ((intrinsicsAssembly == null) &&
+                    if (targetModule == null &&
+                        reader.GetModuleVersionIdOrThrow() == moduleVersionId)
+                    {
+                        targetModule = metadata;
+                    }
+                    if (intrinsicsAssembly == null &&
                         reader.DeclaresType((r, t) => r.IsPublicNonInterfaceType(t, ExpressionCompilerConstants.IntrinsicAssemblyNamespace, ExpressionCompilerConstants.IntrinsicAssemblyTypeName)))
                     {
                         intrinsicsAssembly = identity;
                     }
                 }
-                var reference = MakeAssemblyMetadata(metadata, modulesByName);
-                referencesBuilder.Add(reference);
+            }
+
+            if (kind == MakeAssemblyReferencesKind.AllReferences)
+            {
+                Debug.Assert(referencesBuilder.Count == 0);
+                // CommonReferenceManager<TCompilation, TAssemblySymbol>.Bind()
+                // expects COR library to be included in the explicit assemblies.
+                Debug.Assert(corLibrary != null);
+                if (corLibrary != null && referencesByIdentity.TryGetValue(corLibrary.Name, out var corLibraryReferences))
+                {
+                    referencesBuilder.Add(corLibraryReferences[0].Item2);
+                }
+                Debug.Assert(targetReference != null);
+                if (targetReference != null)
+                {
+                    referencesBuilder.Add(targetReference);
+                }
             }
```

Comment:
There seems to be 2 code paths here that are interwoven but almost entirely distinct. `Kind == AllReferences` and otherwise. Consider refactoring to two different methods or blocks.